bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ReadAheadManager.java

```diff
@@ -0,0 +1,655 @@
+                actualStartLedgerId, actualStartEntryId, location, readAheadTaskExpiredTimeMs));
+        pendingDeletePositions.add(lep);
+    }
+
+    protected ReadAheadTaskStatus getNearestTask(long ledgerId, long entryId) {
+        NavigableMap<Long, ReadAheadTaskStatus> ledgerReadAheadTaskStatuses =
+                inProgressReadAheadTaskStatuses.get(ledgerId);
+        if (ledgerReadAheadTaskStatuses != null) {
+            Map.Entry<Long, ReadAheadTaskStatus> floorEntry = ledgerReadAheadTaskStatuses.floorEntry(entryId);
+            if (floorEntry != null) {
+                return floorEntry.getValue();
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Remove those read-ahead tasks which have already exceeded expired time.
+     * NOTE: this method is NOT thread-safe, thus it should be kept in a single thread.
+     */
+    private void removeExpiredReadAheadTasks() {
+        // cleanup read-ahead pos
+        int reclaimedPositions = 0;
+        while (!pendingDeletePositions.isEmpty()) {
+            ReadAheadPos pos = pendingReadAheadPositions.computeIfPresent(
+                    pendingDeletePositions.peek(),
+                    (lep, rap) -> {
+                        if (rap.isExpired()) {
+                            return null;
+                        }
+                        return rap;
+                    });
+            if (pos == null) {
+                pendingDeletePositions.poll();
+                reclaimedPositions++;
+            } else {
+                break;
+            }
+        }
+
+        // cleanup read-ahead task
+        int reclaimedTasks = 0;
+        while (!pendingDeleteReadAheadTaskStatuses.isEmpty()
+                && pendingDeleteReadAheadTaskStatuses.peek().isExpired()) {
+            ReadAheadTaskStatus readAheadTaskStatus = pendingDeleteReadAheadTaskStatuses.poll();
+            reclaimedTasks++;
+            inProgressReadAheadTaskStatuses.computeIfPresent(
+                    readAheadTaskStatus.ledgerId,
+                    (lid, ledgerReadAheadTaskStatuses) -> {
+                       ledgerReadAheadTaskStatuses.remove(readAheadTaskStatus.startEntryId);
+                       return ledgerReadAheadTaskStatuses.isEmpty() ? null : ledgerReadAheadTaskStatuses;
+                    });
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Pending position map reclaimed {} positions, now is {}. "
+                            + "Read-ahead task map reclaimed {} tasks, now is {}",
+                    reclaimedPositions, pendingDeletePositions.size(),
+                    reclaimedTasks, pendingDeleteReadAheadTaskStatuses.size());
+        }
+    }
+
+    /**
+     * This method could be invoked frequently. Please make it short and simple.
+     */
+    public boolean hitInReadAheadPositions(long ledgerId, long entryId) {
+        AtomicBoolean isHit = new AtomicBoolean(false);
+        pendingReadAheadPositions.computeIfPresent(
```

Comment:
Does pendingReadAheadPositions merge readAhead operation. just like: 
item1: read from (1, 0) to (1, 100)
item2: read from (1, 50) to (1, 150)

item1 will be triggered by read (1, 0) operation, and item2 will be triggered by read (1, 50) operation.